include/swift/Parse/ASTGen.h

```diff
@@ -14,54 +14,107 @@
 #define SWIFT_PARSE_ASTGEN_H
 
 #include "swift/AST/ASTContext.h"
+#include "swift/AST/Decl.h"
 #include "swift/AST/Expr.h"
+#include "swift/Parse/PersistentParserState.h"
 #include "swift/Syntax/SyntaxNodes.h"
 #include "llvm/ADT/DenseMap.h"
 
 namespace swift {
 /// Generates AST nodes from Syntax nodes.
 class ASTGen {
   ASTContext &Context;
-  // A stack of source locations of syntax constructs. Allows us to get the
-  // SourceLoc necessary to create AST nodes for nodes in not-yet-complete
-  // Syntax tree. The topmost item should always correspond to the token/node
-  // that has been parsed/transformed most recently.
-  // todo [gsoc]: remove when possible
-  llvm::SmallVector<SourceLoc, 16> LocStack;
+
+  /// Type cache to prevent multiple transformations of the same syntax node.
+  llvm::DenseMap<syntax::SyntaxNodeId, TypeRepr *> TypeCache;
```

Comment:
Could you explain why the key cannot be `OpaqueSyntaxNode` or `RawSyntax *`?

Reply:
I think it could be either of those.
For me personally, relying on an ID rather than on an address that could change during execution seems more robust, but I'm open to suggestions why pointers would be better solution here.

Reply:
Sorry, never mind about `SyntaxNodeId` or address. 

My concern here is that `RawSyntax` can be reused in the tree.
We need to do this cache for `Expr *`, `Stmt *`, etc. too, right?
For example, for `[ 1, 1]`, these `IntegerLiteralSyntax`s *may* hold the same single instance of `RawSyntax`. But we cannot use single `IntegerLiteralExpr` AST node for multiple position. So I thought we might want to use `std::pair<SyntaxNodeId, SourceLoc>` as the key for this cache table. WDYT?

But, at the same time, I know we are reusing `RawSyntax` only for `TokenSyntax` (via `RawSyntaxTokenCache`). So it shouldn't be a problem for now.